### PR TITLE
New metric classes (#1326)

### PR DESCRIPTION
**Pull request by justusschock**
followup #1326

### CODE REVIEW

File: pytorch_lightning/metrics/__init__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+as a structure to implement your custom metrics for all the fancy research
+you want to do.
+
+For native PyTorch implementations of metrics, it is recommended to use
+the :class:`TensorMetric` which handles automatted DDP syncing and conversions
+to tensors for all inputs and outputs.
+
+If your metrics implementation works on numpy, just use the
+:class:`NumpyMetric`, which handles the automatted conversion of
+inputs to and outputs from numpy as well as automatted ddp syncing.
```
Comment (awaelchli) [inline pytorch_lightning/metrics/__init__.py:17]:
```suggestion
inputs to and outputs from numpy as well as automated ddp syncing.
```

File: pytorch_lightning/metrics/converters.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
+"""
+This file provides functions and decorators for automated input and output
+conversion to/from numpy.ndarray and torch.Tensor as well as utilities to
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:3]:
```suggestion
conversion to/from :class:`numpy.ndarray` and :class:`torch.Tensor` as well as utilities to
```

File: pytorch_lightning/metrics/converters.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    """
+    if isinstance(data, numbers.Number):
+        return torch.tensor([data])
+    # is not array of object
+    elif isinstance(data, np.ndarray) and np_str_obj_array_pattern.search(data.dtype.str) is None:
+        return torch.from_numpy(data)
+    elif isinstance(data, torch.Tensor):
+        return data
+
+    raise TypeError("The given type ('%s') cannot be converted to a tensor!" % type(data).__name__)
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:84]: maybe could use f strings as in rest of code base.

File: pytorch_lightning/metrics/converters.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        return np.array([data])
+    elif isinstance(data, np.ndarray):
+        return data
+
+    raise TypeError("The given type ('%s') cannot be converted to a numpy array!" % type(data).__name__)
+
+
+def _numpy_metric_conversion(func_to_decorate: Callable) -> Callable:
+    """
+    Decorator Handling the argument conversion for metrics working on numpy.
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:108]:
```suggestion
    Decorator handling the argument conversion for metrics working on numpy.
```

File: pytorch_lightning/metrics/converters.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    All inputs of the decorated function will be converted to numpy and all
+    outputs will be converted to Tensors
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:110]:
```suggestion
    outputs will be converted to tensors.
```

File: pytorch_lightning/metrics/converters.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+                           group: Optional[Any] = None,
+                           reduce_op: Optional[torch.distributed.ReduceOp] = None,
+                           ) -> torch.Tensor:
+    """
+    Function to reduce the tensors from several ddp processes to one master process
+
+    Args:
+        result: the value to sync and reduce (typically tensor or number)
+        group: the process group to gather results from. Defaults to all processes (world)
+        reduce_op: the reduction operation. Defaults to sum
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:158]:
```suggestion
        reduce_op: the reduction operation. Defaults to :func:`torch.sum`.
```
not sure if it is torch.sum, but probably?
also appears in other places

File: pytorch_lightning/metrics/converters.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+
+def numpy_metric(group: Optional[Any] = None,
+                 reduce_op: Optional[torch.distributed.ReduceOp] = None) -> Callable:
+    """
+    This decorator shall be used on all function metrics working on numpy arrays.
+
+    It handles the argument conversion and DDP reduction for metrics working on numpy.
+    All inputs of the decorated function will be converted to numpy and all
+    outputs will be converted to Tensors.
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:187]:
```suggestion
    outputs will be converted to tensors.
```

File: pytorch_lightning/metrics/converters.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    return decorator_fn
+
+
+def tensor_metric(group: Optional[Any] = None,
+                  reduce_op: Optional[torch.distributed.ReduceOp] = None) -> Callable:
+    """
+    This decorator shall be used on all function metrics working on tensors.
+
+    It handles the argument conversion and DDP reduction for metrics working on tensors.
+    All inputs and outputs of the decorated function will be converted to tensors .
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:213]:
```suggestion
    All inputs and outputs of the decorated function will be converted to tensors.
```

File: pytorch_lightning/metrics/metric.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+from pytorch_lightning.metrics.converters import tensor_metric, numpy_metric
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
+
+__all__ = ['Metric', 'TensorMetric', 'NumpyMetric']
+
+
+class Metric(DeviceDtypeModuleMixin, torch.nn.Module, ABC):
+    """
+    Abstract Base Class for metric implementation.
```
Comment (awaelchli) [inline pytorch_lightning/metrics/metric.py:16]:
```suggestion
    Abstract base class for metric implementation.
```

File: pytorch_lightning/metrics/metric.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            reduce_op: the operation to perform during reduction within DDP (only needed for DDP training).
+                Defaults to sum.
+        """
+        super().__init__(name)
+        self._orig_call = numpy_metric(group=reduce_group,
+                                       reduce_op=reduce_op)(super().__call__)
+
+    def __call__(self, *args, **kwargs) -> torch.Tensor:
+        def _to_device_dtype(x: torch.Tensor) -> torch.Tensor:
+            return x.to(device=self.device, dtype=self.dtype)
```
Comment (awaelchli) [inline pytorch_lightning/metrics/metric.py:100]: should the metrics also do non blocking transfers maybe? this was recently added in another pr when the batch is moved. or is there too little impact?

File: pytorch_lightning/utilities/apply_func.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
+from collections import Mapping, Sequence
+from typing import Any, Callable, Union
+
+
+def apply_to_collection(data: Any, dtype: Union[type, tuple], function: Callable, *args, **kwargs) -> Any:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:5]: this is very useful!! :heart:

**Comment by Borda** [comment]
@justusschock I would not do this with standard PR as we want to get this merge without squash, so I would prefer to rebase this branch to master and merge it directly so we do not lose track of particular changes...

File: .github/workflows/ci-testing.yml [hunk 12]
```diff
@@ -54,6 +54,11 @@ jobs:
       run: |
         python -c "req = open('requirements.txt').read().replace('torch>=1.1', 'torch<1.5') ; open('requirements.txt', 'w').write(req)"
 
+    # versions <= 1.3 may have issues on mac with some BLAS ops due to missing mkl (https://github.com/pytorch/pytorch/issues/18996)
+    - name: Setup MacOS Minimal
```
Comment (Borda) [inline .github/workflows/ci-testing.yml:58]: need to update README table accordingly

File: pytorch_lightning/metrics/converters.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    """
+    if isinstance(data, numbers.Number):
+        return torch.tensor([data])
+    # is not array of object
+    elif isinstance(data, np.ndarray) and np_str_obj_array_pattern.search(data.dtype.str) is None:
+        return torch.from_numpy(data)
+    elif isinstance(data, torch.Tensor):
+        return data
+
+    raise TypeError("The given type ('%s') cannot be converted to a tensor!" % type(data).__name__)
```
Comment (Borda) [inline pytorch_lightning/metrics/converters.py:84]:
```suggestion
    raise TypeError(f"The given type ('{type(data).__name__}') cannot be converted to a tensor!")
```

File: pytorch_lightning/metrics/metric.py (hunk 10 above)
Comment (Borda) [reply pytorch_lightning/metrics/metric.py:100]: I guess that there was a comment that the non-blocking is by default...

Comment (awaelchli) [reply pytorch_lightning/metrics/metric.py:100]: where did you find that?

Comment (Borda) [reply pytorch_lightning/metrics/metric.py:100]: ok, not found it, so maybe just my bad :]

File: pytorch_lightning/metrics/converters.py (hunk 6 above)
Comment (justusschock) [reply pytorch_lightning/metrics/converters.py]: no, this is a special sync operator, but probably behaves similar to torch.sum

**Comment by justusschock** [comment]
@Borda then let's do it with the rebasing thing. Anything I need to do for this? :)

**Comment by justusschock** [comment]
@awaelchli Thanks, You're a like a human grammar dict :D

**Comment by Borda** [comment]
> @Borda then let's do it with the rebasing thing. Anything I need to do for this? :)

looking at the [commit history](https://github.com/PyTorchLightning/pytorch-lightning/pull/1877/commits) and there was only one PR merged to this branch so we can go the standard way... I ve rebased it so it shall be fine now...

File: pytorch_lightning/metrics/metric.py (hunk 10 above)
Comment (justusschock) [reply pytorch_lightning/metrics/metric.py]: I changed this (@Borda the comment you're referring to was in another context :) )

**Comment by justusschock** [comment]
Thanks :) I added the non-blocking stuff. So it should be ready to go :)

File: pytorch_lightning/metrics/converters.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/metrics/converters.py]: oh yes, now I see it. Jirka already commited my suggestion, so I will make another one so you can revert it, ok? It's only in this one place.

File: pytorch_lightning/metrics/converters.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+                           group: Optional[Any] = None,
+                           reduce_op: Optional[torch.distributed.ReduceOp] = None,
+                           ) -> torch.Tensor:
+    """
+    Function to reduce the tensors from several ddp processes to one master process
+
+    Args:
+        result: the value to sync and reduce (typically tensor or number)
+        group: the process group to gather results from. Defaults to all processes (world)
+        reduce_op: the reduction operation. Defaults to :func:`torch.sum`.
```
Comment (awaelchli) [inline pytorch_lightning/metrics/converters.py:158]:
```suggestion
        reduce_op: the reduction operation. Defaults to sum.
```
In response to [comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/1877#discussion_r427356889)
